pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasuresFunctors.hpp

```diff
@@ -16,27 +16,455 @@
 #include <Kokkos_Core.hpp>
 #include <Kokkos_StdAlgorithms.hpp>
 
+#include "UtilKokkos.hpp"
+
 namespace Pennylane::LightningKokkos::Functors {
+
 /**
  * @brief Compute probability distribution from StateVector.
  *
+ * @tparam PrecisionT StateVector precision.
+ * @tparam DeviceType Kokkos execution space.
  * @param arr_ StateVector data.
- * @param probabilities_ Discrete probability distribution.
+ * @param wires_ Wires for which the probability is computed.
+ * @param all_indices_ Base indices.
+ * @param all_offsets_ Offset indices.
  */
-template <class PrecisionT> struct getProbFunctor {
-    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
-    Kokkos::View<PrecisionT *> probability;
+template <class PrecisionT, class DeviceType> class getProbsFunctor {
+  public:
+    // Required for functor:
+    using execution_space = DeviceType;
+    using value_type = PrecisionT[];
+    const unsigned value_count;
 
-    getProbFunctor(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
-                   Kokkos::View<PrecisionT *> probability_)
-        : arr(arr_), probability(probability_) {}
+    using ComplexT = Kokkos::complex<PrecisionT>;
+    Kokkos::View<ComplexT *> arr;
+    Kokkos::View<std::size_t *> all_indices;
+    Kokkos::View<std::size_t *> all_offsets;
+    getProbsFunctor(const Kokkos::View<ComplexT *> &arr_,
+                    const std::vector<std::size_t> &wires_,
+                    const Kokkos::View<std::size_t *> all_indices_,
+                    const Kokkos::View<std::size_t *> all_offsets_)
+        : value_count{1U << wires_.size()}, arr{arr_},
+          all_indices{all_indices_}, all_offsets{all_offsets_} {}
 
     KOKKOS_INLINE_FUNCTION
-    void operator()(const std::size_t k) const {
-        const PrecisionT REAL = arr(k).real();
-        const PrecisionT IMAG = arr(k).imag();
-        probability(k) = REAL * REAL + IMAG * IMAG;
+    void init(PrecisionT dst[]) const {
+        for (unsigned i = 0; i < value_count; ++i)
```

Comment:
Is there a risk of memory out of bound here? Any reason to apply pre-increment here?

Reply:
No, but I'll change it for style. Thanks for spotting it.